### ISSUE TITLE
impeller: rrect path has dimples

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#156422]
## reproduction steps
See `GradientOvalStrokeMaskBlurSigmaZero` introduced in https://github.com/flutter/engine/pull/55717

## seen results
<img width="467" alt="Screenshot 2024-10-08 at 10 41 22 AM" src="https://github.com/user-attachments/assets/bf649320-dc13-43c7-bed7-731700cf98fa">

## expected results
I'd expect smooth sides, notice the dimples at either ends of the shape.

## version
https://github.com/flutter/engine/tree/772142e52853fcc3779f22bf5f3203010ee253ee

**Comment by jonahwilliams** [comment]
RRect?

**Comment by gaaclarke** [comment]
> RRect?

Yup. https://github.com/flutter/engine/pull/55717/files#diff-de4beabd04cd7bef4aa21a5ee85881eb9df0c336b351bfef29bee69ad31b9a5cR97

**Comment by jonahwilliams** [comment]
Ahh, but stroked rrect so its going through the generic path tessellation code instead of the special rrect tessellation: https://github.com/flutter/engine/blob/main/impeller/display_list/canvas.cc#L676

**Comment by gaaclarke** [comment]
Here's the geometry where the dimples happen

<img width="888" alt="Screenshot 2024-10-14 at 3 23 28 PM" src="https://github.com/user-attachments/assets/7e151d89-9cee-47e6-be03-bf66eb1b871f">

**Comment by jonahwilliams** [comment]
Since this is a stroked path, I suspect the problem has to do with computations of the begin/end vectors

**Comment by jonahwilliams** [comment]
We've had similar bugs in the past

**Comment by gaaclarke** [comment]
The polyline from `CreateTempPolyline` is good in StrokePathGeometry::GetPositionBuffer:
<img width="844" alt="Screenshot 2024-10-14 at 4 41 01 PM" src="https://github.com/user-attachments/assets/9c1a045a-b08b-4b8e-829a-76ae8511cbae">

**Comment by gaaclarke** [comment]
The source of this problem is in `AddVerticesForLinearComponent`.  If you skip the loop if `point_i == 11 || point_y == 33` the output will look almost the same but the dimples will be slightly larger.  Here are the vertices that are getting added for the calls to `AddVerticesForLinearComponent`.  The points for point_i==11 and point_i==33 look suspicious because points are repeated.  Since this is a triangle strip I think it should be zigzagging back and forth around the polyline:

```
[ERROR:flutter/impeller/entity/geometry/stroke_path_geometry.cc(212)] 0 (15, 10)
(15, -10)
(185, 10)
(185, -10)

[ERROR:flutter/impeller/entity/geometry/stroke_path_geometry.cc(212)] 11 (210, 30)
(190, 30)
(210, 30)
(190, 30)

[ERROR:flutter/impeller/entity/geometry/stroke_path_geometry.cc(212)] 22 (185, 50)
(185, 70)
(15, 50)
(15, 70)

[ERROR:flutter/impeller/entity/geometry/stroke_path_geometry.cc(212)] 33 (-10, 30)
(10, 30)
(-10, 30)
(10, 30)
```

**Comment by gaaclarke** [comment]
I think it's because the 2 polylines points are really close:
```
- point_i
11
- polyline.GetPoint(point_i)
{...}
x =
200
y =
30.0000019
- polyline.GetPoint(point_i + 1)
{...}
x =
200
y =
29.9999981
```

**Comment by jonahwilliams** [comment]
Are we using a `==` where we need a `ScalarNearlyEquals` somewhere?

**Comment by jonahwilliams** [comment]
The stroked rrect polyline is going to be generated from our PathBuilder, see https://github.com/flutter/engine/blob/main/impeller/geometry/path_builder.cc#L153

**Comment by gaaclarke** [comment]
The close points are coming from `PathBuilder::AddRoundedRect`, the "Right line" section, `AddLinearComponent` is called with (200,30.0000019), (200, 29.9999981).

**Comment by gaaclarke** [comment]
The following patch fixes the dimples:
```diff
--- a/impeller/geometry/path_builder.cc
+++ b/impeller/geometry/path_builder.cc
@@ -284,6 +284,9 @@ void PathBuilder::AddContourComponent(const Point& destination,
 }
 
 void PathBuilder::AddLinearComponent(const Point& p1, const Point& p2) {
+  if (p1.GetDistance(p2) < 0.01) {
+    return;
+  }
   auto& points = prototype_.points;
   points.push_back(p1);
   points.push_back(p2);
```

<img width="436" alt="Screenshot 2024-10-17 at 10 45 45 AM" src="https://github.com/user-attachments/assets/fc70e428-8864-4ad3-80c6-c4427b912e6c">

**Comment by jonahwilliams** [comment]
If we know this is always the last component we could also insert a close instead of a line to the beginning (I think that is what is happening?)

**Comment by gaaclarke** [comment]
It's not the last component, it's here: https://github.com/flutter/engine/blob/31aaaaad868743b38ac3b7165f0d58eca94e521b/impeller/geometry/path_builder.cc#L180

The question is do we want to figure out this segment just for the rounded rect or across the board.  I think we should try to do it across the board.  PR incoming.

**Comment by flar** [comment]
> The following patch fixes the dimples:
> 
> ```diff
> --- a/impeller/geometry/path_builder.cc
> +++ b/impeller/geometry/path_builder.cc
> @@ -284,6 +284,9 @@ void PathBuilder::AddContourComponent(const Point& destination,
>  }
>  
>  void PathBuilder::AddLinearComponent(const Point& p1, const Point& p2) {
> +  if (p1.GetDistance(p2) < 0.01) {
> +    return;
> +  }
>    auto& points = prototype_.points;
>    points.push_back(p1);
>    points.push_back(p2);
> ```

So, not inserting a path segment prevents a gap? I would have thought it would be the other way around.

Also, is this in device coordinates or local coordinates? .01 is not a predictable size without knowing the transform.

**Comment by gaaclarke** [comment]
> Also, is this in device coordinates or local coordinates? .01 is not a predictable size without knowing the transform.

Yea, that was just a hack to demonstrate the problem.  In the PR I use ScalarNearlyEquals which regardless of the transform we are presumably hitting the acceptable limits of float32.

**Comment by flar** [comment]
I think I can answer my own question - we close contours with a join_proc which assumes that it will be followed by an additional segment after its work is done. In particular, the default miter join proc will assume that it can just omit the miter point and the right thing will happen, but there are no more segments added after it. Even if it doesn't choose to omit itself, it uses the bevel joiner to insert 3 points and then adds a 4th which is the miter, again leaving the pen not at the final position and assuming that the next segment will continue from there accurately - which might or might not happen.

In the case of this being the "closing join" - it definitely will not happen.

Basically, there are a number of assumptions in the join_proc code that don't really work in the overall architecture of the stroke widening code. We should fix these assumptions.

It would appear that all that eliminating short segments does it make the segments before the tightly curved part stop one earlier and so the rejection code in the miter does not omit the actual join.

All join procs need to have a set of assumptions added - such as guaranteeing that they will close the gap regardless of any determination of whether they are "too colinear to really be noticeable".

That will fix this problem...

**Comment by flar** [comment]
More on the failure of the join procs - there is no good consideration of which point was the last point inserted in the outgoing vertices and so they look like they may often overdraw. In particular, a bevel join draws 4 vertices when it probably can get away with 1 if it had any idea of the state the preceding path was left in. But, at least it ensures that the final point it inserts is set up for the new line to start on.

The round and miter joins simply insert nothing leaving the previous point randomly decided by the last inserted contour and the next point randomly decided by the next contour to be inserted. That could work if those were well defined, or if they checked the state (WasLastPointLeftOrRight(), WillNextPointBeLeftOrRight()), but there is no state to check and they have no code to check this.

So, simply returning from a join proc should be outlawed unless there is better state tracking and management so they know how the two segments will be drawn and whether they will line up. If the join is too small to decorate then they should insert a Bevel join instead (that's how the miter limit is defined and you can see that the code to process the miter limit in the miter join proc is after a bevel was already inserted).

And it would look like the way that miter works is to insert the 3 vertices for a bevel and then backtrack to the miter location - "in the hopes that it makes a good launch point for the next segment to insert from" - without any defined constraints or state checking to verify that.

Finally, bevel shouldn't need to insert 3 points to draw itself. Potentially 1 or 0 points would be enough depending on how well defined the outlining state is defined or can be queried...

**Comment by flar** [comment]
Suggested fix for now: Call `CreateBevelAndGetDirection` in the early return cases for both Miter and Round joins.

Suggested fix in the longer term: Design the state machine of the outline generation better and have the join procs use it to insert optimal joins.

**Comment by gaaclarke** [comment]
> I ask again - how is eliminating a line segment fixing a gap? It sounds like the problem is somewhere else and this is fixing a symptom case rather than a problem.

It demonstrably eliminates the gap.  How it works is that the code that is making the triangle strip around the polyline, `AddVerticesForLinearComponent`, is just generating vertices that overlap with one another.

> It sounds like the problem is somewhere else and this is fixing a symptom case rather than a problem.

Possibly there is a problem in the joining logic where having overlapping vertices is goofing it up.  But we don't want overlapping vertices in the first place.  So if the joining logic works when the vertices are the most efficiently represented, I don't think it's behaving incorrectly.  Said another way, if we fix any joining logic to accommodate bad input, we are just masking bad input.  The linked PR filters out that bad input.

**Comment by flar** [comment]
I created a gist to try to reproduce this with larger segments that are nearly colinear, but didn't see any gaps forming:

https://gist.github.com/flar/18753764f8ea5f024a172b75ab1bac3d
